### PR TITLE
feat: Expose space ID

### PR DESCRIPTION
**Pull request by dmaretskyi**
### Motivation / Background

This Pull Request ...

### CODE REVIEW

File: packages/common/keys/src/public-key.test.ts [hunk 1]
```diff
@@ -60,4 +60,11 @@ describe('PublicKey', () => {
     const key = PublicKey.random();
     expect(PublicKey.equals(key, PublicKey.from(key.toHex()))).to.be.true;
   });
+
+  test('base32', () => {
+    const key = PublicKey.randomOfLength(20); // space keys will be cut to first 20 bytes of sha-256 hash
```
Comment (richburdon) [inline packages/common/keys/src/public-key.test.ts:65]: `// A comment is a sentence.`

File: packages/common/keys/src/public-key.ts [hunk 2]
```diff
@@ -183,6 +197,10 @@ export class PublicKey implements Equatable {
     return this.asBuffer().toString('hex');
   }
 
+  toMultibase32(): string {
+    return 'B' + base32Encode(this._value, 'RFC4648');
+  }
+
   truncate(length = undefined) {
```
Comment (richburdon) [inline packages/common/keys/src/public-key.ts:204]: `length = 8` default?

File: packages/common/keys/src/space-id.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import base32Encode from 'base32-encode';
+
+import { invariant } from '@dxos/invariant';
+
+import { randomBytes } from './random-bytes';
+
+/**
+ * A unique identifier for a space.
+ */
+export type SpaceId = string & { __SpaceId: true };
```
Comment (richburdon) [inline packages/common/keys/src/space-id.ts:15]: make __SypaceID a symbol (and everywhere else we need this pattern?)

File: packages/common/keys/src/space-id.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+export const SpaceId = Object.freeze({
```
Comment (richburdon) [inline packages/common/keys/src/space-id.ts:17]: move some or all of the comments from your GH issue to the jsdoc here?

File: packages/common/keys/src/space-id.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  },
+  isValid: (value: string): value is SpaceId => {
+    return typeof value === 'string' && value.startsWith(MULTIBASE_PREFIX) && value.length === ENCODED_LENGTH;
+  },
+  random: (): SpaceId => {
+    return SpaceId.encode(randomBytes(SpaceId.byteLength));
+  },
+});
+
+const MULTIBASE_PREFIX = 'B';
```
Comment (richburdon) [inline packages/common/keys/src/space-id.ts:38]: when/where is the IPFS multiformat prefix used?

File: packages/core/echo/echo-pipeline/src/space/space.ts [hunk 6]
```diff
@@ -199,3 +209,22 @@ export class Space extends Resource {
     log('closed');
   }
 }
+
+/**
+ * Space keys are generated by creating a keypair, and then taking the first 20 bytes of the SHA-256 hash of the public key and encoding them to multibase RFC4648 base-32 format (prefixed with B, see Multibase Table).
+ * Inspired by how ethereum addresses are derived.
+ */
+export const deriveIdFromSpaceKey = async (spaceKey: PublicKey): Promise<SpaceId> => {
```
Comment (richburdon) [inline packages/core/echo/echo-pipeline/src/space/space.ts:217]: createIdFromSpaceKey

File: packages/sdk/client/src/echo/space-proxy.ts [hunk 7]
```diff
@@ -118,6 +118,10 @@ export class SpaceProxy implements Space {
     this._membersUpdate.emit(_data.members ?? []);
   }
 
+  get id(): SpaceId {
+    return this._data.id as SpaceId;
```
Comment (richburdon) [inline packages/sdk/client/src/echo/space-proxy.ts:122]: is `id` ambiguous?

**Review by richburdon — APPROVED** [review]

File: packages/common/keys/src/space-id.ts (hunk 3 above)
Comment (dmaretskyi) [reply packages/common/keys/src/space-id.ts:17]: I think a string tag is simpler and more then enough for this usecase. automerge uses them - https://github.com/automerge/automerge-repo/blob/main/packages/automerge-repo/src/types.ts#L5

I'm not really sure how symbols would behave here

File: packages/common/keys/src/public-key.ts (hunk 2 above)
Comment (dmaretskyi) [reply packages/common/keys/src/public-key.ts:204]: separate PR

**Pull request merged**